src/Analyzer/InDepthQueryTreeVisitor.h

```diff
@@ -84,4 +84,47 @@ class InDepthQueryTreeVisitor
 template <typename Derived>
 using ConstInDepthQueryTreeVisitor = InDepthQueryTreeVisitor<Derived, true /*const_visitor*/>;
 
+/** Visitor that use another visitor to visit node only if condition for visiting node is true.
+  * For example, your visitor need to visit only specific query tree nodes.
+  *
+  * Condition interface:
+  * struct Condition
+  * {
+  *     bool operator()(VisitQueryTreeNodeType & node)
+  *     {
+  *         return shouldNestedVisitorVisitNode(node);
+  *     }
+  * }
+  */
+template <typename Visitor, typename Condition, bool const_visitor = false>
+class InDepthQueryTreeConditionalVisitor : public InDepthQueryTreeVisitor<InDepthQueryTreeConditionalVisitor<Visitor, Condition, const_visitor>, const_visitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitor<InDepthQueryTreeConditionalVisitor<Visitor, Condition, const_visitor>, const_visitor>;
+    using VisitQueryTreeNodeType = typename Base::VisitQueryTreeNodeType;
+
+    explicit InDepthQueryTreeConditionalVisitor(Visitor & visitor_, Condition & condition_)
+        : visitor(visitor_)
+        , condition(condition_)
+    {
+    }
+
+    bool shouldTraverseTopToBottom() const
+    {
+        return visitor.shouldTraverseTopToBottom();
+    }
+
+    void visitImpl(VisitQueryTreeNodeType & query_tree_node)
+    {
+        if (condition(query_tree_node))
+            visitor.visit(query_tree_node);
```

Comment:
I guess we don't want it to traverse a subtree. It will result in `O(N^2)` complexity.
```suggestion
            visitor.visitImpl(query_tree_node);
```

Reply:
We want to run visitor only on some nodes that are visited by this conditional visitor. For example we want to run visitor for each subquery. Conditional visitor will iterate all nodes. And for subquery nodes it will call nested visitor. Nested visitor will not go into subqueries. So in result we will iterate over each node only 2 times. If nested visitor or condition is wrong, it can result in quadratic complexity, but I am not sure if this class will be used a lot. Better to keep it until we finish migration, if no one uses it we can delete it.